src/libraries/System.Runtime.Numerics/src/System/Numerics/BigNumber.cs

```diff
@@ -494,35 +494,243 @@ private static bool HexNumberToBigInteger(ref BigNumberBuffer number, out BigInt
+                    {
+                        int remainingIntDigitCount = Math.Max(numberScale - totalDigitCount, 0);
+                        ReadOnlySpan<char> intDigitsSpan = chunkDigits.Slice(0, Math.Min(remainingIntDigitCount, chunkDigits.Length));
+
+                        bool endReached = false;
+
+                        // Storing these captured variables in locals for faster access in the loop.
+                        uint _partialValue = partialValue;
+                        int _partialDigitCount = partialDigitCount;
+                        int _totalDigitCount = totalDigitCount;
+
+                        for (int i = 0; i < intDigitsSpan.Length; i++)
+                        {
+                            char digitChar = chunkDigits[i];
+                            if (digitChar == '\0')
+                            {
+                                endReached = true;
+                                break;
+                            }
+
+                            _partialValue = _partialValue * 10 + (uint)(digitChar - '0');
+                            _partialDigitCount++;
+                            _totalDigitCount++;
+
+                            // Update the buffer when enough partial digits have been accumulated.
+                            if (_partialDigitCount == MaxPartialDigits)
+                            {
+                                MultiplyAdd(ref currentBuffer, TenPowMaxPartial, _partialValue);
+                                _partialValue = 0;
+                                _partialDigitCount = 0;
+                            }
+                        }
+
+                        // Check for nonzero digits after the decimal point.
+                        if (!endReached)
+                        {
+                            ReadOnlySpan<char> fracDigitsSpan = chunkDigits.Slice(intDigitsSpan.Length);
+                            for (int i = 0; i < fracDigitsSpan.Length; i++)
+                            {
+                                char digitChar = fracDigitsSpan[i];
+                                if (digitChar == '\0')
+                                {
+                                    break;
+                                }
+                                if (digitChar != '0')
+                                {
+                                    return false;
+                                }
+                            }
+                        }
+
+                        partialValue = _partialValue;
+                        partialDigitCount = _partialDigitCount;
+                        totalDigitCount = _totalDigitCount;
+
+                        return true;
+                    }
+                }
+                else
                 {
-                    if (!ProcessChunk(digitsChunk.Span, ref currentBuffer))
+                    if (numberScale < 0)
```

Comment:
Can we actually get a negative `numberScale` here? Parse should treat `AllowDecimalPoint` and `AllowExponent` as invalid.

Reply:
`AllowDecimalPoint` and `AllowExponent` sometimes become valid. so I think this check is necessary.
You can execute code like below:
```cs
Console.WriteLine(BigInteger.Parse("1.2345e+10", NumberStyles.AllowDecimalPoint | NumberStyles.AllowExponent)); // 12345000000
Console.WriteLine(BigInteger.Parse("1.2345e-10", NumberStyles.AllowDecimalPoint | NumberStyles.AllowExponent)); // raise System.FormatException: 'The value could not be parsed.'
```
I also place assert for `0 <= numberScale` here, then it failed.

Reply:
Ah, right. I apparently wasn't awake enough when I was reading the code this morning.

I misread `InvalidNumberStyles` and didn't see the `~`, coupled with not expect decimal points to be allowed given we don't allow it for other integers.

This seems correct/fine then.